san/extras/event_study.py

```diff
@@ -0,0 +1,286 @@
+    pyplot.ylabel("% Cumulative Return")
+    pyplot.xlabel("Time Window")
+    pyplot.grid(b=None, which=u'major', axis=u'y')
+    pyplot.legend()
+
+def plot_cumulative_abnormal_returns(returns, abnormal_returns, x_ticks):
+    pyplot.figure(figsize=FIGURE_SIZE)
+
+    returns.plot(xticks=x_ticks, label="Average Cumulative Returns")
+    abnormal_returns.plot(xticks=x_ticks, label="Abnormal Average Cumulative Returns")
+
+    pyplot.axhline(
+        y=abnormal_returns.loc[0],
+        linestyle='--',
+        color='black',
+        alpha=.3,
+        label='Drift'
+    )
+
+    pyplot.axhline(
+        y=abnormal_returns.max(),
+        linestyle='--',
+        color='black',
+        alpha=.3
+    )
+
+    pyplot.title("Cumulative Abnormal Returns versus Cumulative Returns")
+    pyplot.ylabel("% Cumulative Return")
+    pyplot.xlabel("Time Window")
+    pyplot.grid(b=None, which=u'major', axis=u'y')
+    pyplot.legend()
+
+
+def plot_cumulative_return_with_errors(returns, std_devs, sample_size):
+    """
+    Plotting the same graph but with error bars
+    """
+    pyplot.figure(figsize=FIGURE_SIZE)
+
+    pyplot.errorbar(returns.index, returns, xerr=0, yerr=std_devs, label=f"N={sample_size}")
+    pyplot.grid(b=None, which=u'major', axis=u'y')
+    pyplot.title("Cumulative Return from Events with error")
+    pyplot.xlabel("Window Length (t)")
+    pyplot.ylabel("Cumulative Return (r)")
+    pyplot.legend()
+    pyplot.show()
+
+def plot_abnormal_cumulative_return_with_errors(abnormal_volatility, abnormal_returns, sample_size):
+    """
+    Capturing volatility of abnormal returns
+    """
+    pyplot.figure(figsize=FIGURE_SIZE)
+
+    pyplot.errorbar(
+        abnormal_returns.index,
+        abnormal_returns,
+        xerr=0,
+        yerr=abnormal_volatility,
+        label=f"N={sample_size}"
+    )
+
+    pyplot.grid(b=None, which=u'major', axis=u'y')
+    pyplot.title("Abnormal Cumulative Return from Events with error")
+    pyplot.xlabel("Window Length (t)")
+    pyplot.ylabel("Cumulative Return (r)")
+    pyplot.legend()
+    pyplot.show()
+
+def build_day_numbers(starting_point):
+    """
+    Create our range of day_numbers that will be used to calculate returns
+    Looking from -starting_point to +starting_point to create timeframe band
+    """
+    return [i for i in range(-starting_point, starting_point)]
+
+def event_study(data, ev_data, starting_point=30, benchmark='bitcoin', origin_zero=True):
+    if ev_data.symbol[ev_data.symbol == 'bitcoin'].count()!=0:
+         benchmark='ethereum'
+
+    all_returns = {}
+    all_std_devs = {}
+    all_benchmark_returns = {}
+    all_abnormal_returns = {}
+    abnormal_volatility = {}
+    total_sample_size = {}
+
+    day_numbers = build_day_numbers(starting_point)
+
+    for day_num in day_numbers:
+        returns = []
+        benchmark_returns = []
+        abnormal_returns = []
+        sample_size = 0
+
+        for date, row in ev_data.iterrows():
```

Comment:
The whole body of this for (or its parent for?) should probably be a separate function. Here we have nested `if`s and `for`s which is not readable/maintainable

Reply:
Yeah. This function is way too long. Let's try to break it down into logical pieces.

Reply:
Alright. Ricardo changed it to be separate functions. Have a look.